language/ankush-playground/src/Vote.move

```diff
@@ -0,0 +1,260 @@
+        assert(num_votes_required > 0, EINVALID_APPROVAL_POLICY);
+
+        // Using a nested loop verify that allowed_voters does
+        // not contain duplicates
+        let i = 0;
+        let len = Vector::length(&allowed_voters);
+        while (i < len) {
+            let e1 = Vector::borrow(&allowed_voters, i);
+            let j = i + 1;
+            while (j < len) {
+                let e2 = Vector::borrow(&allowed_voters, j);
+                assert(e1 != e2, EINVALID_APPROVAL_POLICY);
+                j = j + 1;
+            };
+            i = i + 1;
+        };
+
+        let policy = ApprovalPolicy {
+            allowed_voters,
+            num_votes_required,
+        };
+
+        let approval_policies = &mut borrow_global_mut<ApprovalPolicies>(@DiemRoot).approval_policies;
+
+        let i = 0;
+        let len = Vector::length(approval_policies);
+        let found = false;
+        while (i < len) {
+            let approval_policy_map = Vector::borrow_mut(approval_policies, i);
+            if (&approval_policy_map.proposal_type == &proposal_type) {
+                // update the approval policy
+                approval_policy_map.policy = *(&policy);
+                found = true;
+                break
+            };
+            i = i + 1;
+        };
+
+        if (!found) {
+            Vector::push_back(approval_policies, ApprovalPolicyMap {
+                proposal_type,
+                policy,
+            });
+        };
+
+    }
+
+    fun get_approval_policy(
+        proposal_type: vector<u8>,
+        addr: address,
+    ): Option<ApprovalPolicy> acquires ApprovalPolicies {
+        let approval_policies = &borrow_global<ApprovalPolicies>(addr).approval_policies;
+
+        let i = 0;
+        let len = Vector::length(approval_policies);
+        while (i < len) {
+            let approval_policy_map = Vector::borrow(approval_policies, i);
+            if (&approval_policy_map.proposal_type == &proposal_type) {
+                return Option::some(*(&approval_policy_map.policy))
+            };
+            i = i + 1;
+        };
+        Option::none()
+    }
+
+    public fun construct_ballot_id(
+        counter: u64,
+        proposer: address,
+    ): BallotID {
+        BallotID {
+            counter,
+            proposer,
+        }
+    }
+
+    public fun create_ballot<Proposal: store>(
+        ballot_account: &signer,
+        proposal: Proposal,
+        proposal_type: vector<u8>,
+        start_timestamp_secs: u64,
+        expiration_timestamp_secs: u64
+    ): BallotID acquires ApprovalPolicies, Ballots, BallotCounter {
+        assert(start_timestamp_secs <= expiration_timestamp_secs, EINVALID_BALLOT_TIMESTAMPS);
+        let ballot_address = Signer::address_of(ballot_account);
+        if (!exists<BallotCounter>(ballot_address)) {
+            move_to(ballot_account, BallotCounter {
+                counter: 0,
+            });
+        };
```

Comment:
Yup. The reason I did this is to avoid using the root signer as much as possible in this code. Initializing the global counter under the root module would require the root signer.